runtime/major_gc.c

```diff
@@ -2102,8 +2184,16 @@ int caml_init_major_gc(caml_domain_state* d) {
                   caml_addrmap_iterator(&d->mark_stack->compressed_stack);
 
   /* Fresh domains do not need to performing marking or sweeping. */
-  d->sweeping_done = 1;
-  d->marking_done = 1;
+  if (caml_gc_phase == Phase_sweep_main) {
+    d->sweeping_done = 1;
+    d->marking_done = 0;
+    atomic_fetch_add(&num_domains_to_mark, 1);
```

Comment:
It is not immediately clear to me why these counters are incremented. Can you provide an intuition, please?

Reply:
In trunk, new domains are always spawned in a state where marking and sweeping have completed. With markdelay, this strategy is not always valid: in `Phase_sweep_main`, marking cannot have completed because it has not yet begun.

So, in `Phase_sweep_main`, we start new domains in a state where sweeping is done but marking is not done. The invariant of `num_domains_to_mark` is that it is always equal to the number of running domains with `marking_done == 0`, so we need to increment it in this case. (Likewise for `ephe_cycle_info.num_domains_todo`).

Reply:
I defer to @stedolan here, but the following is my understanding:
Before this change, the fresh domain doesn't need to mark or sweep in the current collection cycle, because we snapshot-at-the-beginning, so anything reachable at the start of the cycle will be marked and swept by the other domains, and this fresh domain allocates `MARKED`.
With this change, all domains (including the fresh domain) allocate `UNMARKED` until marking begins (`caml_allocation_status()`), so this fresh domain needs to mark, both its own roots (which may get pointers to `UNMARKED` blocks on its own or other major heaps) and its own mark-stack.

Reply:
I'm adding some comments here.